display_list/display_list_unittests.cc

```diff
@@ -3781,5 +3797,81 @@ TEST_F(DisplayListTest, SaveLayerBoundsClipDetectionSimpleClippedRect) {
   EXPECT_TRUE(expector.all_bounds_checked());
 }
 
+class DepthExpector : public virtual DlOpReceiver,
+                      virtual IgnoreAttributeDispatchHelper,
+                      virtual IgnoreTransformDispatchHelper,
+                      virtual IgnoreClipDispatchHelper,
+                      virtual IgnoreDrawDispatchHelper {
+ public:
+  explicit DepthExpector(std::vector<uint32_t> expectations)
+      : depth_expectations_(std::move(expectations)) {}
+
+  void save() override {
+    // This method should not be called since we override the variant with
+    // the max_content_depth parameter.
+    FAIL() << "save(no depth parameter) method should not be called";
+  }
+
+  void save(uint32_t max_content_depth) override {
+    ASSERT_LT(index_, depth_expectations_.size());
+    EXPECT_EQ(depth_expectations_[index_], max_content_depth)
+        << "at index " << index_;
+    index_++;
+  }
+
+  void saveLayer(const SkRect& bounds,
+                 SaveLayerOptions options,
+                 const DlImageFilter* backdrop) override {
+    // This method should not be called since we override the variant with
+    // the max_content_depth parameter.
+    FAIL() << "saveLayer(no depth parameter) method should not be called";
+  }
+
+  void saveLayer(const SkRect& bounds,
+                 const SaveLayerOptions& options,
+                 uint32_t max_content_depth,
+                 const DlImageFilter* backdrop) override {
+    ASSERT_LT(index_, depth_expectations_.size());
+    EXPECT_EQ(depth_expectations_[index_], max_content_depth)
+        << "at index " << index_;
+    index_++;
+  }
+
+ private:
+  size_t index_ = 0;
+  std::vector<uint32_t> depth_expectations_;
+};
+
+TEST_F(DisplayListTest, SaveContentDepthTest) {
+  DisplayListBuilder builder;
+  builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 1
+
+  builder.Save();  // lasts through depth 6
+  {
+    builder.Translate(5, 5);
+    builder.DrawRect({10, 10, 20, 20}, DlPaint());  // depth 2
+
+    builder.SaveLayer(nullptr, nullptr);  // lasts through depth 4
```

Comment:
Ah, this is something I should have specified. SaveLayers have a draw call call that also needs a unique depth. But they are drawn _after_ all of their children have been rendered, so the depth we use for them must be greater than all contained children drawn within the SaveLayers.

I recommend [replicating this test](https://github.com/flutter/engine/blob/main/impeller/aiks/aiks_unittests.cc#L2929-L2974) against the DL depth values to ensure that:
1. SaveLayers are reserved a depth value, and
2. clips interact correctly with SaveLayers.

Reply:
I noticed that while I was updating dl_dispatcher.cc...

The odd thing is that they only allocate the depth on the restore after recording the max_depth, right? The clips inside them will be rendered to the "restore copy - 1" max depth and then the restore copy happens above that depth so it is only clipped by the enclosing clips.

Reply:
I think I've accounted for this in the latest commit.